typing/printtyp.ml

```diff
@@ -2176,51 +2218,96 @@ let unification_error env tr txt1 ppf txt2 ty_expect_explanation =
       print_labels := true;
       raise exn
 
-let report_unification_error ppf env tr
-    ?(type_expected_explanation = fun _ -> ())
-    txt1 txt2 =
-  wrap_printing_env env (fun () -> unification_error env tr txt1 ppf txt2
-                            type_expected_explanation)
+let error trace_format env tr txt1 ppf txt2 ty_expect_explanation =
+  let prepare_expansion_head empty_tr = function
+    | Errortrace.Diff d -> Some (Errortrace.map_diff (may_prepare_expansion empty_tr) d)
+    | _ -> None
+  in
+  reset ();
+  let tr = prepare_trace trace_format (fun t t' -> t, hide_variant_name t') tr in
+  let mis = mismatch trace_format txt1 env tr in
+  handle_trace (filter_trace trace_format) prepare_expansion_head
+    env tr trace_format.incompatibility_phrase mis txt1 ppf txt2 ty_expect_explanation
+
+let report_error trace_format ppf env tr
+      ?(type_expected_explanation = fun _ -> ())
```

Comment:
I'm not sure how to avoid it – what did you have in mind?

Reply:
A simple option would be to have a separate function for unification error report with an expected type explanation since the `report_error` function is mostly glue code. However, I am not sure if this is a good idea or not.

Reply:
I decided to go with this change, or at least a variant of it – we now expose `report_unification_error`, `report_equality_error`, and `report_moregen_error` as separate functions, with the unification case having the optional argument and the latter two not having it.  As a side effect, I also no longer expose the `trace_format` type.